### ISSUE TITLE
no-useless-concat should ignore cases where at least one of the concatenated strings contains `

### ISSUE THREAD

**Issue by finnigantime** [eslint/eslint#13426]
no-useless-concat should ignore cases where at least one of the concatenated strings contains "`" (and possibly "${" as well).

Escape characters can always be used to exclude backticks and `${`, so for example to write this string without concatenation:

```js
'[`REDIS_DB`, `${REDIS_DBS.' + `${this.serviceNames.serviceNameSnake.toUpperCase()}` + '}`]',
```

we could do:

```js
`[\`REDIS_DB\`, \`\${REDIS_DBS.${this.serviceNames.serviceNameSnake.toUpperCase()}}\`]`,
```

but that's awkward and difficult to read in my opinion - it's difficult to read because it's hard to tell which backtick characters and `${` get inserted into the final string. This seems like a legit use case for string concatenation. It's fairly common in codegen to insert ` and `${`. Why not allow the author to opt for string concatenation in such cases to allow for more readable code without all the escape characters?

**What rule do you want to change?**
`no-useless-concat`

**Does this change cause the rule to produce more or fewer warnings?**
fewer

**How will the change be implemented? (New option, new default behavior, etc.)?**
My preference would be a new default behavior. Could also add an option.

**Please provide some example code that this change will affect:**

```js
      '[`REDIS_DB`, `${REDIS_DBS.' + `${this.serviceNames.serviceNameSnake.toUpperCase()}` + '}`]',
```

**What does the rule currently do for this code?**
emits a warning

```
error Unexpected string concatenation of literals. [no-useless-concat]
  103 |
  104 |     this.addEnvMap([
> 105 |       '[`REDIS_DB`, `${REDIS_DBS.' + `${this.serviceNames.serviceNameSnake.toUpperCase()}` + '}`]',
      |                                    ^
  106 |       '[`REDIS_HOST`, ``]',
  107 |       '[`REDIS_PORT`, `${wiring.redisPort(def)}`]',
```

**What will the rule do after it's changed?**
no warnings

**Are you willing to submit a pull request to implement this change?**
not at this time

**Comment by mysticatea** [comment]
Thank you for your proposal.

Out of curiosity, why do you use `` `${this.serviceNames.serviceNameSnake.toUpperCase()}` `` instead of `this.serviceNames.serviceNameSnake.toUpperCase()`? That rule warns the concatenation of two string literals, so it will not warn it if you don't wrap `this.serviceNames.serviceNameSnake.toUpperCase()`.

**Comment by finnigantime** [comment]
@mysticatea you're absolutely correct, in this example the backticks and '${}' can go away. I don't know how I didn't notice that when setting up eslint, I must have glossed over and not looked at this line in enough detail. Thanks, I'll close this ticket!
